cmd/skuba/cluster/check.go

```diff
@@ -0,0 +1,133 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package cluster
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubeadm"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/pkg/skuba"
+	"github.com/SUSE/skuba/pkg/skuba/actions/cluster/upgrade"
+	"github.com/rikatz/kubepug/lib"
+	"github.com/rikatz/kubepug/pkg/formatter"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/klog"
+)
+
+type checkOptions struct {
+	K8sVersion      string
+	ForceDownload   bool
+	APIWalk         bool
+	SwaggerDir      string
+	ShowDescription bool
+}
+
+var (
+	format    string
+	filename  string
+	inputFile string
+)
+
+// newUpgradeCheckCmd creates a new `skuba check` cobra command
+func newUpgradeCheckCmd() *cobra.Command {
+	checkOptions := &checkOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "check --kubernetes-version=<version> --swaggerDir=<directory> --api-walk=<true|fasle>",
+		Short: "Print Upgrade Check information",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			kubernetesConfigFlags := genericclioptions.NewConfigFlags(true)
+			kubeAdminFile := skuba.KubeConfigAdminFile()
+			kubernetesConfigFlags.KubeConfig = &kubeAdminFile
+
+			config := lib.Config{
+				K8sVersion:      checkOptions.K8sVersion,
+				ForceDownload:   checkOptions.ForceDownload,
+				APIWalk:         checkOptions.APIWalk,
+				SwaggerDir:      checkOptions.SwaggerDir,
+				ShowDescription: checkOptions.ShowDescription,
+				ConfigFlags:     kubernetesConfigFlags,
+				Input:           inputFile,
+			}
+
+			if config.K8sVersion == "" {
+				client, err := kubernetes.GetAdminClientSet()
+				if err != nil {
+					return err
+				}
+				currentClusterVersion, err := kubeadm.GetCurrentClusterVersion(client)
+				if err != nil {
+					return err
+				}
+				availableVersions := kubernetes.AvailableVersions()
+				upgradePath, err := upgrade.CalculateUpgradePath(currentClusterVersion, availableVersions)
+				if err != nil {
+					return err
+				}
+
+				var nextClusterVersion *version.Version
+				if len(upgradePath) > 0 {
+					nextClusterVersion = upgradePath[0]
+				} else {
+					klog.Warning("Already on the latest version, nothing to check.\nFor a specific version use the `kubernetes-version` flag.")
+					return nil
+				}
+				config.K8sVersion = fmt.Sprintf("v%s", nextClusterVersion.String())
+			}
```

Comment:
`else` shouldn't we ensure the specified version is something known to kubepug? I'm just concerned a user would use an older version of skuba to perform a `check` for a version of kubernetes that didn't exist when the `skuba` binary got created.

Reply:
I don't think this would matter, `skuba` would fail is the version of k8s is not a valid upstream version.
If the version is specified by the user, `kubepug` will find the correct swagger file for that one; if it does not exist it will fail.

Reply:
I see, I didn't know about the swagger usage at the time of writing this review